Feature/matrix

kerenGuez opened the pull request.

kerenGuez committed:
My solution for the matrix exercise

kerenGuez committed:
matrix solution

kerenGuez committed:
shorter lines

natsala13 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,158 @@
+            tuple: New tuple of tuples, matrix, main diagonal is 1, the rest 0.
+        """
+        unity_matrix = []
+        for row_num, row in enumerate(range(size)):
```
Using enumerate is great also here it have no sense since row and row_num will be identical.

natsala13 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,158 @@
+                else:
+                    tuple_nums.append(0)
+
+            unity_matrix.append(tuple(tuple_nums))
```
Replace it with a smart list comprehension

natsala13 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,158 @@
+        Returns:
+            tuple: Tuple of tuples filled with ones.
+        """
+        ones_mat = [tuple([1 for num in range(size)]) for i in range(size)]
```
Try and use list concatenation for creating list of ones.

natsala13 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,158 @@
+        return Matrix(tuple((tuple((x + y for x, y in zip(tuple1, tuple2))))
+                      for tuple1, tuple2 in zip(self.tuples, other)))
+
+    def __sub__(self, other):
```
Reuse already implemented methods if you can.

natsala13 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,158 @@
+        """Return the regular multiplication when the order is reversed."""
+        return self * other
+
+    def __truediv__(self, other):
```
Reuse already implemented methods if you can.

natsala13 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,158 @@
+        """
+        result = [tuple(sum(n_a * n_b for n_a, n_b in zip(a_row, b_col))
+                        for b_col in zip(*sec_mat))
+                  for a_row in self.tuples]
```
Excellent !!!

natsala13 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,158 @@
+    def _multiply_by_scalar(self, scalar):
+        """Return new matrix multiplied by scalar."""
+        return Matrix(tuple((tuple((num * scalar for num in each_tuple))
+                             for each_tuple in self.tuples)))
```
Don't need self.tuples
Why? (understand it before changing).

natsala13 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,158 @@
+            return self._multiplication(other.tuples)
+
+        else:
+            raise TypeError("Second parameter must be int/matrix.")
```
I think argument would fir here better.
Who is the second? will __mul__ be called only when a matrix is at the left?

natsala13 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,158 @@
+    def __add__(self, other):
+        """Return result of 2 matrices addition."""
+        return Matrix(tuple((tuple((x + y for x, y in zip(tuple1, tuple2))))
+                      for tuple1, tuple2 in zip(self.tuples, other)))
```
Correct indentation.

natsala13 replied:
Way to much brackets.

kerenGuez committed:
changes were committed.
-used list comprehension for unity method
-used string concatenation for ones method
-changed self.tuples to self where needed
-removed useless checks in __mul__ method
-reused already implemented methods in stead of my older __truediv__ and __sub__ ones
-corrected indentation in __add__ method

kerenGuez committed:
shortened line

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+class Matrix:
```
python file should start with doc line.

rfire01 replied:
Add unittests

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+        self._is_tuple()
+        self._is_even()
+
+    def __init__(self, matrix_tuples):
```
init should be the first function in class

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+        Attributes:
+            __matrix (tuple): a tuple of tuples representing a matrix.
+    """
+    def _is_even(self):
```
The name of the function make it sound that the function return true\false if its even

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+        """Check that matrix is made of a tuple of tuples."""
+        if isinstance(self.__matrix, tuple):
+            for mat_tuple in self.__matrix:
+                if not isinstance(mat_tuple, tuple):
```
can you think about a way to do it with list comprehension?

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+        Returns:
+            tuple: New tuple of tuples, matrix, main diagonal is 1, the rest 0.
+        """
+        lis = [tuple(1 if col == row else 0 for col in range(size))
```
Why not make it a tuple from the start instead of list?

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+        """Calculate the multiplication of 2 matrices.
+
+            Args:
+                sec_mat (tuple): A matrix to multiply with.
```
give full names

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+        """
+        result = [tuple(sum(n_a * n_b for n_a, n_b in zip(a_row, b_col))
+                        for b_col in zip(*sec_mat))
+                  for a_row in self]
```
Very good!
just save the conversion from list to tuple

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+        return tuple(result)
+
+    def _multiply_by_scalar(self, scalar):
+        """Return new matrix multiplied by scalar."""
```
missing doc

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+
+    def _multiply_by_scalar(self, scalar):
+        """Return new matrix multiplied by scalar."""
+        return Matrix(tuple((tuple((num * scalar for num in each_tuple))
```
Too many parenthesis

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+            Matrix: The result matrix.
+        """
+        if isinstance(other, Matrix):
+            return self._multiplication(other.tuples)
```
blank line

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+        return Matrix(tuple((tuple((num * scalar for num in each_tuple))
+                             for each_tuple in self)))
+
+    def __mul__(self, other):
```
Will this function always work? for any two matrices?

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+        """
+        if isinstance(other, Matrix):
+            return self._multiplication(other.tuples)
+        return self._multiply_by_scalar(other)
```
what will happen if I do Matrix() * [1,2,3]?

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+
+    def __rmul__(self, other):
+        """Return the regular multiplication when order is reversed"""
+        return self * other
```
This is not correct

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+    def __truediv__(self, other):
+        """Return new matrix divided by scalar."""
+        if other == 0:
+            raise ZeroDivisionError("Cannot divide by 0.")
```
blank line

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+
+    def __add__(self, other):
+        """Return result of 2 matrices addition."""
+        add_result = ((tuple((x + y for x, y in zip(tuple1, tuple2))))
+                      for tuple1, tuple2 in zip(self, other))
```
too many parenthesis

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+
+    def __eq__(self, other):
+        """Check for equality between matrices."""
+        return self.tuples == other
```
does it work?

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,139 @@
+
+    def __ne__(self, other):
+        """Check if the two matrices are not equal."""
+        return self.tuples != other
```
same

kerenGuez committed:
I corrected issues in matrix file and created a unit test for it.

kerenGuez committed:
shortened lines

kerenGuez committed:
shorter line

kerenGuez committed:
removed unnecessary lines

natsala13 commented on advanced/matrix/matrix.py:
```diff
@@ -72,78 +64,106 @@ def ones(cls, size):
     def __ne__(self, other):
         """Check if the two matrices are not equal."""
-        return self.tuples != other
+        return self.tuples != other.tuples
```
Try reusing already implemented method.

natsala13 commented on advanced/matrix/test_matrix.py:
```diff
@@ -0,0 +1,120 @@
+
+        with self.assertRaises(TypeError):
+            Matrix("hi")
+            Matrix(((1, 2, 3), [4, 5, 6], 5))
```
This line is unreachable

natsala13 commented on advanced/matrix/test_matrix.py:
```diff
@@ -0,0 +1,120 @@
+    def test_initialize(self):
+        """Test initialization of matrix."""
+        matrix_tuples = ((2, 5), (8, 3))
+        my_matrix = Matrix(matrix_tuples)
```
You have already initialized a matrix in setup, why creating another one?

natsala13 commented on advanced/matrix/test_matrix.py:
```diff
@@ -0,0 +1,120 @@
+
+    def test_iter(self):
+        """Test the iterator of the matrix."""
+        lines = ""
+        for line in self.mat_a:
+            lines += f"{line}\n"
+        self.assertEqual(lines, "(1, 2)\n(3, 4)\n")
```
Instead of using string just check every line of the matrix

natsala13 commented on advanced/matrix/test_matrix.py:
```diff
@@ -0,0 +1,120 @@
+
+    def test_hash(self):
+        """Test if hash method."""
+        dictionary = {Matrix(((1, 1), (2, 2))): 2,
```
You already initialized matrices in setup method.

natsala13 commented on advanced/matrix/test_matrix.py:
```diff
@@ -0,0 +1,120 @@
+"""Unit testing for the Matrix class."""
```
**Change all methods so that every test case would check only one case. (one assert). If you want to check a few test cases of add_matrix for example, then write few different unit test methods checking every case separately.**

kerenGuez committed:
applied changes necessary.

kerenGuez committed:
shorter lines

kerenGuez committed:
minor changes

natsala13 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,178 @@
+        result = tuple(tuple(sum(n_a * n_b for n_a, n_b in zip(a_row, b_col))
+                       for b_col in zip(*other_matrix))
+                       for a_row in self)
+        return Matrix(result)
```
If one want to create a new class inheriting from this one, and try to use this method from the inheriting class. What would happen?

Anyway, try not using Hard Coded class name.
We are looking for a code which require a minimum amount of work for every adaptation we want to add to it.

natsala13 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,178 @@
+"""An implementation of the class for the data type 'Matrix'."""
```
Great Work!

kerenGuez committed:
changed hard coded class name instances

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,178 @@
+    def _check_tuple(self):
+        """Check that matrix is made of a tuple of tuples."""
+        if isinstance(self.__matrix, tuple):
+            if not all(isinstance(mat_tuple, tuple)
+                       for mat_tuple in self.__matrix):
```
You should use naming that will be as clear as possible for other people who don't know your code.
For example, instead of mat_tuple, maybe matrix_row (to make it clear what this tuple means)

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,178 @@
+            if not all(isinstance(mat_tuple, tuple)
+                       for mat_tuple in self.__matrix):
+
+                raise TypeError("Object is not a tuple's matrix.")
```
Object is not tuple of tuples

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,178 @@
+
+        Returns:
+            Matrix : The new matrix multiplied by scalar.
+
```
remove

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,178 @@
+        ones_mat = (((1,) * size,) * size)
+        return cls(ones_mat)
+
+    def _multiplication(self, other_matrix):
```
if you have several multiplication, be specific in the function name

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,178 @@
+        Returns:
+            Matrix: The result matrix.
+        """
+        try:
+            self._check_valid_matrix(other)
+            return self._multiplication(other.tuples)
+
+        except TypeError:
```
There is no need for try-except here

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,178 @@
+                raise ValueError("Cannot operate on different sized matrices.")
+
+        else:
+            raise TypeError("Cannot operate on different sized matrices.")
```
This doesn't seem correct

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,178 @@
+
+    def __rmul__(self, other):
+        """Return the regular multiplication when in reversed order."""
+        return self * other
```
If other is matrix this is not correct.
Can you tell me why even so, this answer is still correct?

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,178 @@
+        """Return the regular multiplication when in reversed order."""
+        return self * other
+
+    def __truediv__(self, other):
```
Fix in all functions:
When there is an option that you will raise an exception in function, you need in the function doc add after the "returns" a section of "Raises" and what can it raise (you can look at the coding style doc how it looks)

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,178 @@
+    def __add__(self, other):
+        """Return result of 2 matrices addition."""
+        self._check_valid_matrix(other)
+        add_result = (tuple(x + y for x, y in zip(tuple1, tuple2))
+                      for tuple1, tuple2 in zip(self, other))
```
same, better naming

rfire01 commented on advanced/matrix/test_matrix.py:
```diff
@@ -0,0 +1,148 @@
+        with self.assertRaises(ValueError):
+            Matrix(((1, 2, 3), (1, 2, 3)))
+
+    def test_unity_2_dimensional(self):
+        """Test unity method on 2 by 2 matrix."""
+        self.assertEqual(Matrix.unity(2), Matrix(((1, 0), (0, 1))))
+
+    def test_unity_3_dimensional(self):
+        """Test unity method on 3 by 3 matrix."""
+        self.assertEqual(Matrix.unity(3),
+                         Matrix(((1, 0, 0), (0, 1, 0), (0, 0, 1))))
```
You may missed it, but you checked here two functions:
1. unity function
2. __eq__ function
Generally it is ok to use multiple functions in single test, but you need to check they are working before using them (the functions that are not the purpose of the test). 
In this case the test might have failed, but because of the __eq__ and not because the unity
(This is correct for all tests that do it)

rfire01 commented on advanced/matrix/test_matrix.py:
```diff
@@ -0,0 +1,148 @@
+        """Test multiplication by float."""
+        self.assertEqual(self.mat_a * 0.9, Matrix(((0.9, 1.8), (2.7, 3.6))))
+
+    def test_multiplication_of_matrices(self):
+        """Test multiplication of matrices."""
+        self.assertEqual(self.mat_a * self.mat_b, Matrix(((15, 21), (33, 47))))
+
+    def test_multiplication_reversed(self):
```
what about matrix mul reversed?

rfire01 commented on advanced/matrix/test_matrix.py:
```diff
@@ -0,0 +1,148 @@
+
+    def test_addition(self):
+        """Test the method for addition of 2 matrices."""
+        self.assertEqual(self.mat_a + Matrix.unity(2),
```
why not used mat_b if you already created it?

rfire01 commented on advanced/matrix/test_matrix.py:
```diff
@@ -0,0 +1,148 @@
+
+    def test_iter(self):
+        """Test the iterator of the matrix."""
+        mat_a_iter = iter(self.mat_a)
+        self.assertEqual(next(mat_a_iter), (1, 2))
+
+        self.assertEqual(next(mat_a_iter), (3, 4))
```
instead of using next and splitting to 2 asserts, you can do something like list(mat_a) or tuple(mat_a) etc., and then compare the resulted list\tuple. 
list and tuple know how to iterate on iterable variable

rfire01 commented on advanced/matrix/test_matrix.py:
```diff
@@ -0,0 +1,148 @@
+
+        self.assertEqual(next(mat_a_iter), (3, 4))
+
+    def test_eq(self):
+        """Test the equality check between matrices."""
+        self.assertEqual(self.mat_a == self.mat_b, False)
```
This is not enough.
This test shows that comparing two matrices will return False, however, this test will pass even if the __eq__ function was
def __eq__(...):
   return False
When testing you need to make sure all of its functionality works

kerenGuez committed:
applied necessary changes

kerenGuez committed:
shorter line

kerenGuez committed:
shorter lines

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,220 @@
+            matrix_rows (tuple): a tuple of tuples representing a matrix.
+
+        Raises:
+            TypeError: self is not a tuple of tuples.
+            ValueError: self doesn't have an equal number of rows and columns.
```
self doesn't exists yet at this point (you should be explicit what are you talking about)

rfire01 commented on advanced/matrix/matrix.py:
```diff
@@ -0,0 +1,220 @@
+    def __truediv__(self, other):
+        """Return new matrix divided by scalar.
+        Args:
+            other (float/int): Scalar to divide by.
```
in those cases you can just write number as the type

rfire01 commented on advanced/matrix/test_matrix.py:
```diff
@@ -0,0 +1,166 @@
+        with self.assertRaises(TypeError):
+            Matrix(((1, 2, 3), [4, 5, 6], 5))
+
+    def test_value_validation(self):
```
maybe test_invalid_constructor or something like this.
its not clear from the name what you are testing

rfire01 commented on advanced/matrix/test_matrix.py:
```diff
@@ -0,0 +1,166 @@
+        self.assertEqual(str(self.mat_a), "((1, 2), (3, 4))")
+
+
+if __name__ == "__main__":
```
missing test for tuples property

kerenGuez committed:
small changes
